### ISSUE TITLE
Capture both stdout and stderr with Data.Conduit.Process?

### ISSUE THREAD

**Issue by joehillen** [snoyberg/conduit#250]
I keep finding myself is situations were I need to capture both stdout and stderr in parallel.

I've implemented it 2 different ways. Both "work", but both have problems.

First is this way:

```haskell
sourceProcessWithConsumers :: CreateProcess
                           -> Consumer ByteString IO () -- stdout
                           -> Consumer ByteString IO () -- stderr
                           -> IO ExitCode
sourceProcessWithConsumers cp consumerStdout consumerStderr = do
    ( ClosedStream, (sourceStdout, closeStdout)
                  , (sourceStderr, closeStderr), cph) <- streamingProcess cp
    race_ (sourceStdout $$ consumerStdout)
          (sourceStderr $$ consumerStderr)
    closeStdout
    closeStderr
    waitForStreamingProcess cph
```

which seems to pause between each process run. I assume it's blocking on either `closeStdout` or `closeStderr`.

The second way is:

```haskell
sourceProcessWithConsumers :: CreateProcess
                           -> Consumer ByteString IO () -- stdout
                           -> Consumer ByteString IO () -- stderr
                           -> IO ExitCode
sourceProcessWithConsumers cp consumerStdout consumerStderr = do
    ( ClosedStream, (sourceStdout, closeStdout)
                  , (sourceStderr, closeStderr), cph) <- streamingProcess cp
    race_ (sourceStdout $$ consumerStdout >> closeStdout)
          (sourceStderr $$ consumerStderr >> closeStderr)
    waitForStreamingProcess cph
```

This one is fast but it causes a file descriptor leak that is evident when I try to spawn 100+ concurrent processes.

Does anyone know the *right* way to do this? My knowledge of conduits is fairly limited. I would love to have a way to capture both stderr and stdout concurrently and add it to the library.

**Comment by snoyberg** [comment]
Are you compiling your application with `-threaded`?

**Comment by joehillen** [comment]
Turns out my slow down was something completely unrelated.

I actually get the same speed with both versions, so the first one if obviously the correct one.

Sorry for wasting your time.

But it would be nice to have a this functionality in `conduit-extra`. Without it, I had to copy this code block into my application, which is inconvenient. 

```
instance (r ~ (), MonadIO m, o ~ ByteString) => OutputSink (ConduitM i o m r) where
    osStdStream = (\(Just h) -> return $ sourceHandle h, Just CreatePipe)
instance (r ~ (), r' ~ (), MonadIO m, MonadIO n, o ~ ByteString) => OutputSink (ConduitM i o m r, n r') where
    osStdStream = (\(Just h) -> return (sourceHandle h, liftIO $ hClose h), Just CreatePipe)
```

I can also see myself needing to send data to `stdin`, so something like would be nice too.

```
sourceProcessWithConsumers :: CreateProcess
                           -> Producer IO ByteString -- stdin
                           -> Consumer ByteString IO () -- stdout
                           -> Consumer ByteString IO () -- stderr
                           -> IO ExitCode
```

Would you be interested in a PR?

**Comment by snoyberg** [comment]
I'm not sure about that first code block, but I'm open to including `sourceProcessWithConsumers`. I'm worried though that it's not general enough; something like the following may be better:

```haskell
sourceProcessWithConsumers :: CreateProcess
                           -> Producer IO ByteString -- stdin
                           -> Consumer ByteString IO a -- stdout
                           -> Consumer ByteString IO b -- stderr
                           -> IO (ExitCode, a, b)
```

**Comment by joehillen** [comment]
Good idea.

One question though is the name. "withConsumers" would not be entirely accurate because of the stdin Producer. I can't think of anything good. Any ideas?

**Comment by snoyberg** [comment]
Maybe `WithStreams`?

**Comment by snoyberg** [comment]
PR merged, closing.
